src/theory/quantifiers/sygus/sygus_grammar_cons.cpp

```diff
@@ -822,8 +874,221 @@ void CegGrammarConstructor::mkSygusDefaultGrammar(
+          // as constructed above. It may also be an operator from another
+          // theory that has both an arithmetic return type and an arithmetic
+          // argument (e.g. str.indexof). In either case, we ignore it for the
+          // sake of well-foundedness.
+          isBuiltinArithOp = true;
+          break;
+        }
+        else if (atype != unres_bt)
+        {
+          // It is an external type. This is the case of an operator of another
+          // theory whose return type is arithmetic, e.g. select.
+          hasExternalType = false;
+        }
+      }
+      if (!isBuiltinArithOp)
+      {
+        useConstructor[k] = true;
+        if (hasExternalType)
+        {
+          // If we have an external term in the sum, e.g. select(A,i), we
+          // cannot use a fixed polynomial template.
+          Trace("sygus-grammar-def")
+              << "Cannot use polynomial grammar due to " << sop << std::endl;
+          polynomialGrammar = false;
+        }
+      }
+    }
+    Trace("sygus-grammar-def")
+        << "Done look at operators, num = " << sdti.getNumConstructors()
+        << "..." << std::endl;
+    // we have now decided whether we will use sum-of-monomials or polynomial
+    // Now, extract the terms and set up the polynomial
+    std::vector<Node> sumChildren;
+    std::vector<TypeNode> cargsAnyTerm;
+    std::vector<Node> lambdaVars;
+    for (unsigned k = 0, ncons = sdti.getNumConstructors(); k < ncons; k++)
+    {
+      Trace("sygus-grammar-def") << "Process #" << k << std::endl;
+      if (useConstructor.find(k) == useConstructor.end())
+      {
+        Trace("sygus-grammar-def") << "Skip variable #" << k << std::endl;
+        // builtin operator, as computed above, we skip
+        continue;
+      }
+      const SygusDatatypeConstructor& sdc = sdti.getConstructor(k);
+      Node sop = sdc.d_op;
+      Trace("sygus-grammar-def")
+          << "Monomial variable: #" << k << ": " << sop << std::endl;
+      unsigned nargs = sdc.d_argTypes.size();
+      std::vector<TypeNode> opCArgs;
+      std::vector<Node> opLArgs;
+      if (nargs > 0)
+      {
+        // Take its arguments. For example, if we are building a polynomial
+        // over str.len(s), then our any term constructor would include an
+        // argument of string type, e.g.:
+        //   (lambda s : String, c1, c2 : Int. c1*len(s) + c2)
+        for (unsigned j = 0; j < nargs; j++)
+        {
+          // this is already corresponds to the correct sygus datatype type
+          TypeNode atype = sdc.d_argTypes[j];
+          opCArgs.push_back(atype);
+          // get the builtin type
+          TypeNode btype = sygus_to_builtin[atype];
+          opLArgs.push_back(nm->mkBoundVar(btype));
+        }
+        // Do beta reduction on the operator so that its arguments match the
+        // fresh variables of the lambda we are constructing.
+        sop = datatypes::utils::mkSygusTerm(sop, opLArgs);
```

Comment:
Do you mean that `sop` is a lambda? Isn't it coming directly from the datatype? Overall this part is hard to follow